src/shogun/kernel/LinearARDKernel.cpp

```diff
@@ -47,99 +49,186 @@ CLinearARDKernel::~CLinearARDKernel()
 
 bool CLinearARDKernel::init(CFeatures* l, CFeatures* r)
 {
+	cleanup();
 	CDotKernel::init(l, r);
-
-	init_ft_weights();
-
+	int32_t dim=((CDotFeatures*) l)->get_dim_feature_space();
+	if (m_ARD_type==KT_FULL)
+	{
+		REQUIRE(m_weights.num_cols==dim, "Dimension mismatch between features (%d) and weights (%d)\n",
+			dim, m_weights.num_cols);
+	}
+	else if (m_ARD_type==KT_DIAG)
+	{
+		REQUIRE(m_weights.num_rows==dim, "Dimension mismatch between features (%d) and weights (%d)\n",
+			dim, m_weights.num_rows);
+	}
 	return init_normalizer();
 }
 
-void CLinearARDKernel::init_ft_weights()
-{
-	if (!lhs || !rhs)
-		return;
-
-	int32_t alen, blen;
-
-	alen=((CDenseFeatures<float64_t>*) lhs)->get_num_features();
-	blen=((CDenseFeatures<float64_t>*) rhs)->get_num_features();
 
-	REQUIRE(alen==blen, "Number of right and left hand features must be the "
-			"same\n")
+SGMatrix<float64_t> CLinearARDKernel::compute_right_product(SGVector<float64_t>right_vec,
+	float64_t & scalar_weight)
+{
+	SGMatrix<float64_t> right;
 
-	if (m_weights.vlen != alen)
+	if (m_ARD_type==KT_SCALAR)
 	{
-		m_weights=SGVector<float64_t>(alen);
-		m_weights.set_const(1.0);
+		right=SGMatrix<float64_t>(right_vec.vector,right_vec.vlen,1,false);
+		scalar_weight*=m_weights[0];
 	}
+	else 
+	{
+		right=SGMatrix<float64_t> (m_weights.num_rows,1);
 
-	SG_DEBUG("Initialized weights for LinearARDKernel (%p).\n", this)
+		SGMatrix<float64_t> rtmp(right_vec.vector,right_vec.vlen,1,false);
+
+		if(m_ARD_type==KT_DIAG)
+		{
+			linalg::elementwise_product(m_weights, rtmp, right);
+		}
+		else if(m_ARD_type==KT_FULL)
+		{
+			linalg::matrix_product(m_weights, rtmp, right);
+		}
+		else
+		{
+			SG_ERROR("Unsupported ARD kernel\n");
+		}
+	}
+	return right;
 }
 
-void CLinearARDKernel::set_weight(float64_t w, index_t i)
+float64_t CLinearARDKernel::compute_helper(SGVector<float64_t> avec, SGVector<float64_t>bvec)
 {
-	if (i >= m_weights.vlen)
+	SGMatrix<float64_t> left;
+
+	float64_t scalar_weight=1.0;
+	if (m_ARD_type==KT_SCALAR)
 	{
-		SG_ERROR("Index %i out of range for LinearARDKernel."\
-				 "Number of features is %i.\n", i, m_weights.vlen);
+		left=SGMatrix<float64_t>(avec.vector,1,avec.vlen,false);
+		scalar_weight=m_weights[0];
 	}
+	else 
+	{
+		left=SGMatrix<float64_t>(1,m_weights.num_rows);
 
-	m_weights[i]=w;
-}
+		SGMatrix<float64_t> ltmp(avec.vector,avec.vlen,1,false);
 
-float64_t CLinearARDKernel::get_weight(index_t i)
-{
-	if (i >= m_weights.vlen)
-	{
-		SG_ERROR("Index %i out of range for LinearARDKernel."\
-				 "Number of features is %i.\n", i, m_weights.vlen);
+		SGMatrix<float64_t> left_transpose(left.matrix,left.num_cols,1,false);
+		if(m_ARD_type==KT_DIAG)
+		{
```

Comment:
we dont use { } if the statement is only one line. But thats minor here

i.e.
```
else if(m_ARD_type==KT_FULL)
    linalg::matrix_product(m_weights, ltmp, left_transpose);
```
rather than
```
else if(m_ARD_type==KT_FULL)
		{
			linalg::matrix_product(m_weights, ltmp, left_transpose);
		}
```